src/retry.rs

```diff
@@ -0,0 +1,133 @@
+use crate::QldbError;
+use async_trait::async_trait;
+use rand::thread_rng;
+use rand::Rng;
+use rusoto_core::RusotoError;
+use rusoto_qldb_session::*;
+use std::{cmp::min, time::Duration};
+use tokio::time::delay_for;
+
+pub fn default_retry_policy() -> impl TransactionRetryPolicy {
+    ExponentialBackoffJitterTransactionRetryPolicy::default()
+}
+
+pub fn never() -> impl TransactionRetryPolicy {
+    NeverRetryPolicy {}
+}
+
+/// A retry policy receives an `error` and the `attempt_number` and returns true/false based on whether the driver should retry or not. The function [`on_err`] is marked `async` to allow
+/// implementations to model features such as backoff-jitter without blocking the runtime.
+#[async_trait]
+pub trait TransactionRetryPolicy {
+    async fn on_err(&self, error: &QldbError, attempt_number: u32) -> bool;
+}
+
+/// Don't try this at home.
+pub struct NeverRetryPolicy {}
+
+#[async_trait]
+impl TransactionRetryPolicy for NeverRetryPolicy {
+    async fn on_err(&self, _error: &QldbError, _attempt_number: u32) -> bool {
+        false
+    }
+}
+
+/// https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+pub struct ExponentialBackoffJitterTransactionRetryPolicy {
+    base: u32,
+    cap: u32,
+    max_attempts: u32,
+}
+
+impl ExponentialBackoffJitterTransactionRetryPolicy {
+    pub fn new(
+        base: u32,
+        cap: u32,
+        max_attempts: u32,
+    ) -> ExponentialBackoffJitterTransactionRetryPolicy {
+        ExponentialBackoffJitterTransactionRetryPolicy {
+            base,
+            cap,
+            max_attempts,
+        }
+    }
+}
+
+impl Default for ExponentialBackoffJitterTransactionRetryPolicy {
+    fn default() -> ExponentialBackoffJitterTransactionRetryPolicy {
+        ExponentialBackoffJitterTransactionRetryPolicy::new(10, 5000, 4)
+    }
+}
+
+#[async_trait]
+impl TransactionRetryPolicy for ExponentialBackoffJitterTransactionRetryPolicy {
+    async fn on_err(&self, error: &QldbError, attempt_number: u32) -> bool {
+        match error {
+            QldbError::Rusoto(e) => {
+                let should_retry = match &e {
```

Comment:
Note this list is simply `false` except for `OccConflict`. I need to update it!

Reply:
I left a fixme. I won't fix all the retries in this commit because I really want tests along with the fix.